### ISSUE TITLE
csc.exe compiler error -2146232797 with ref struct C# 7.2 modifier

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#24776]
Hello, both VS 15.5 Preview 5 and the official VS 15.5 fail to compile my library using the C# 7.2 ref struct modifier in one of the structs I'm using.

Here's the link to the exact code that reproduces the error, just clone and try to compile the library &quot;NeuralNetwork.NET&quot;:
https://github.com/Sergio0694/NeuralNetwork.NET/tree/VS_test_csc-compiler-error

Removing the ref struct modifier from the &quot;FloatSpan2D&quot; struct solves the issue.

If this is not a bug and that struct is actually being moved to the heap somehow/somewhere, it'd be useful to have some sort of warning to be able to fix the code error.
Otherwise, I hope this will help you pinpoint the cause of the issue.
Cheers!

_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/160134/cscexe-compiler-error-2146232797-with-ref-struct-c.html
VSTS ticketId: 533142_
_These are the original issue comments:_
(no comments)
_These are the original issue solutions:_
(no solutions)

---

Edit by @OmarTawfik
Smaller repro:
```csharp
public ref struct Tensor {
    public void TryFree() { }
}
public class C {
        public static unsafe void TryFree(params Tensor[] tensors)
        {
            fixed (Tensor* p = tensors)
                for (int i = 0; i < tensors.Length; i++)
                    p[i].TryFree();
        }
}
```
It hits the following assert:
```
System.InvalidOperationException: Unexpected value 'PointerElementAccess expression of Tensor type' of type 'System.String'
   at Microsoft.CodeAnalysis.CSharp.Binder.GetValEscape(BoundExpression expr, UInt32 scopeOfTheContainingExpression)
   at Microsoft.CodeAnalysis.CSharp.Binder.CheckInvocationArgMixing(SyntaxNode syntax, Symbol symbol, BoundExpression receiverOpt, ImmutableArray`1 parameters, ImmutableArray`1 argsOpt, ImmutableArray`1 argRefKindsOpt, ImmutableArray`1 argsToParamsOpt, UInt32 scopeOfTheContainingExpression, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindInvocationExpressionContinued(SyntaxNode node, SyntaxNode expression, String methodName, OverloadResolutionResult`1 result, AnalyzedArguments analyzedArguments, MethodGroup methodGroup, NamedTypeSymbol delegateTypeOpt, DiagnosticBag diagnostics, CSharpSyntaxNode queryClause)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindMethodGroupInvocation(SyntaxNode syntax, SyntaxNode expression, String methodName, BoundMethodGroup methodGroup, AnalyzedArguments analyzedArguments, DiagnosticBag diagnostics, CSharpSyntaxNode queryClause, Boolean allowUnexpandedForm)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindInvocationExpression(SyntaxNode node, SyntaxNode expression, String methodName, BoundExpression boundExpression, AnalyzedArguments analyzedArguments, DiagnosticBag diagnostics, CSharpSyntaxNode queryClause, Boolean allowUnexpandedForm)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindInvocationExpression(InvocationExpressionSyntax node, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindExpressionInternal(ExpressionSyntax node, DiagnosticBag diagnostics, Boolean invoked, Boolean indexed)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindExpression(ExpressionSyntax node, DiagnosticBag diagnostics, Boolean invoked, Boolean indexed)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindExpressionStatement(CSharpSyntaxNode node, ExpressionSyntax syntax, Boolean allowsAnyExpression, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindExpressionStatement(ExpressionStatementSyntax node, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindStatement(StatementSyntax node, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindPossibleEmbeddedStatement(StatementSyntax node, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.ForLoopBinder.BindForParts(ForStatementSyntax node, Binder originalBinder, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.ForLoopBinder.BindForParts(DiagnosticBag diagnostics, Binder originalBinder)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindStatement(StatementSyntax node, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindPossibleEmbeddedStatement(StatementSyntax node, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindFixedStatementParts(FixedStatementSyntax node, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindFixedStatement(FixedStatementSyntax node, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindStatement(StatementSyntax node, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindBlockParts(BlockSyntax node, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.BindMethodBody(MethodSymbol method, TypeCompilationState compilationState, DiagnosticBag diagnostics, ImportChain& importChain, Boolean& originalBodyNested)
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileMethod(MethodSymbol methodSymbol, Int32 methodOrdinal, ProcessedFieldInitializers& processedInitializers, SynthesizedSubmissionFields previousSubmissionFields, TypeCompilationState compilationState)
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileNamedType(NamedTypeSymbol containingType)
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.<>c__DisplayClass21_0.<CompileNamedTypeAsTask>b__0()

```

**Comment by jcouv** [comment]
Tagging @VSadov @OmarTawfik

**Comment by jaredpar** [comment]
Looks like the link from the initial bug report is broken. @Sergio0694 can you point us to a commit where this problem repos?

**Comment by Sergio0694** [comment]
@jaredpar My bad, I removed that branch a while back as I got no answer from my issue and thought the error was already in your backlog, I'll see if I have a copy of that test branch somewhere, or if I can reproduce the same issue in another branch 👍

**Comment by jaredpar** [comment]
@Sergio0694 yeah sorry for the slow response. Ideally VS feedback issues should make it to the right team quickly but it can get delayed at times for a variety of reasons. If you can get the branch please lte us know cause we'd like to make sure that's fixed.

**Comment by Sergio0694** [comment]
@jaredpar No worries, I should have just left that branch there 😊
I'm currently recreating a repro for you to use, and in the meantime I've found another related bug.

**Steps to reproduce**:
- Open VS (I'm using VS Enterprise 15.5.5)
- Clone this: [https://github.com/Sergio0694/NeuralNetwork.NET](https://github.com/Sergio0694/NeuralNetwork.NET)
- Inside the `NeuralNetwork.NET` project, go to APIs > Structs > `Tensor`, and add the `ref` modifier (**Note:** I know the `ref` modifier is not valid for this struct, see the note at the bottom)
- Save the file, and VS will crash with this message (tried twice in a row just now):

![image](https://user-images.githubusercontent.com/10199417/36168394-3d3dc210-10f9-11e8-83b7-2c34c546b6b7.png)

I'll let you know as soon as I have a branch published with a repro for that other bug, if I manage to reproduce it.

**NOTE:** I have changed the library since my original issue, so I can no longer apply the `ref` modifier in theory right now, as I have a few places where I'm storing instances of this `Tensor` struct in the heap. I'll see if I can reproduce that same compiler error anyways.

**Comment by Sergio0694** [comment]
@jaredpar I've got a partial repro here - it doesn't reproduce that same error, but for some reason every single time I open it Visual Studio doesn't last a single minute before crashing.

Hope this helps! 😄

[NeuralNetworkDotNetReproV1.zip](https://github.com/dotnet/roslyn/files/1721322/NeuralNetworkDotNetReproV1.zip)

**Comment by OmarTawfik** [comment]
Seeing if I can get a smaller repro now.

**Comment by OmarTawfik** [comment]
Attached the root cause to the issue (please see the top message). It is an unexpected value in escape rules.
cc @VSadov

**Comment by Sergio0694** [comment]
@OmarTawfik Sure, glad I could help! 😄

**Comment by OmarTawfik** [comment]
I'll take this one.

**Comment by jcouv** [comment]
Omar fixed this in 15.7 (https://github.com/dotnet/roslyn/pull/24909)
